File: LeetCode_520/main.cpp
/**
    LeetCode_520
    https://leetcode.com/problems/detect-capital/description/
    MatrixDeity, 2018.
*/

#include <iostream>
#include <string>
#include <algorithm>

using namespace std;

class Solution {
public:
    bool detectCapitalUse(string word) {
        int cap = count_if(word.begin(), word.end(), ::isupper);
        return cap == 1 && isupper(word[0]) || cap == word.size() || cap == 0;
    }
};

int main() {
    return 0;
}

File: LeetCode_520/main.go
/**
  LeetCode_520
  https://leetcode.com/problems/detect-capital
  MatrixDeity, 2023.
*/

package leetcode520

import "unicode"

func detectCapitalUse(word string) bool {
	isCapitalized := true
	isTitled := true
	isLowered := true

	for i, char := range word {
		isLowered = isLowered && unicode.IsLower(char)
		isCapitalized = isCapitalized && unicode.IsUpper(char)
		isTitled = isTitled && (i == 0 && unicode.IsUpper(char) || i > 0 && unicode.IsLower(char))
	}

	return isCapitalized || isLowered || isTitled
}
